### PR TITLE
[4.0] Added additional TinyMCE 5 configuration option for "external_plugins"

### PR DESCRIPTION
**Pull request by martyf**
### Summary of Changes
Added support for Tiny's "external_plugins" configuration option. This has been done using a repeatable pairing of plugin names and paths. 

External Plugins allow developers to use their own custom plugins, that live outside of Tiny's source folders, to help improve the content author's experience through adding of new tasks or custom workflows. They really open up the potential for developers to more easily integrate their own plugins with Joomla 4, especially given the power of what TinyMCE 5 can do.

Using this option is 100% optional. There is no loss of functionality if this configuration option is not set by the administrator.

### Testing Instructions
Build and install Joomla from the source as per usual.

When editing the TinyMCE Editor Plugin, a new field is available at the bottom of the main "Plugin" screen. This allows external plugins to be added at a per-set basis.

If used, the plugin's button can be added (albeit crudely) to the toolbar by using the Custom Button option.

You can get the source for a trivial plugin at:
https://github.com/martyf/tiny-plugin-helloworld

Steps to implement are:

1. Place the plugin JS file somewhere (could vary based on how the dev wants to set it all up - for example, "/tiny-plugins/helloworld.js"
2. Go to the Joomla 4 Admin, and TinyMCE Plugin settings, and select the Set (such as 0 for Admin). Enter a new External Plugin (bottom):
- **Name:** helloworld
- **Path:** /tiny-plugins/helloworld.js
3. In "custom buttons", enter the plugin button - in this case, `helloworld`.
4. Create an Article, and you should see a "Hello Button" in the Tiny toolbar
5. Click the button to trigger the plugin - this is a totally trivial example that demonstrates what Tiny 5 can do with its UI Components library (new with v5) - select an option, and on selecting it, will insert the markup. Totally trivial, but is an example plugin.

It is also worth checking that when you have no external plugins configured, the editing process still works. You can:
1. Remove the External Plugin configuration from the TinyMCE Plugin settings, and remove from "Custom Buttons", then Save and Close.
2. Create/Edit an Article, and the editor loads without error or warning.

### Documentation Changes Required

[TinyMCE editor plugin documentation](https://help.joomla.org/proxy?keyref=Help40:Extensions_Plugin_Manager_Edit&lang=en#Editor_-_TinyMCE "https://help.joomla.org/proxy?keyref=Help40:Extensions_Plugin_Manager_Edit&lang=en#Editor_-_TinyMCE") needs to be updated.

(added by @richard67)

### CODE REVIEW

**Comment by brianteeman** [comment]
Thanks for creating this.

My 10c personal opinion is that we should not accept this PR on principal. It is providing direct support for a commercial product that is not GPL (or compatible). In addition the terms of the subscription agreement for tiny.cloud are very restrictive of my freedom to use the software. (see https://about.tiny.cloud/legal/tiny-cloud-services-subscription-agreement/)

There is nothing wrong with this being an addon listed on the JED but for me it goes against our core values.

**Comment by martyf** [comment]
Not necessarily - even the cloud version of Tiny provides the latest build (where new free features are added) where as the included version needs to be updated separately. There is no cost to use an API key with Tiny Cloud - just for premium plugins, if wanted. But I do get where you're coming from too.

What about for external plugin support? This would really open up potential to help do more with the core editor in J4.

Happy to do a separate PR for just that change if you like.

**Comment by brianteeman** [comment]
@martyf no cost is not the same as GPL (or compatible). Your Terms ((see https://about.tiny.cloud/legal/tiny-cloud-services-subscription-agreement/) are explicit that it is not opensource and that it restricts my usage.

As for the custom plugins - dont we already have that possibility

**Comment by martyf** [comment]
I’m not from Tiny, so not my terms. I’m just a developer who uses their product. 

The current external plugins don’t give enough flexibility for plugins in different locations. These need a path to load correctly as it gives greater flexibility with where to store the plugin source within the folder structure. The “Externals Plugins” feature here resolves that.

**Comment by brianteeman** [comment]
> I’m not from Tiny, so not my terms. I’m just a developer who uses their product.

Sorry I put two and two together from your location and made five

> The current external plugins don’t give enough flexibility for plugins in different locations

Can you give a concrete example please

**Comment by martyf** [comment]
Basically if I make my own plugin (to do anything) I need to load an external JS file. This is done using Tiny’s “external_plugins” option. 

The “plugins” option, which is currently used, looks in the Tiny plugins directory - so basically plugins that are core to Tiny. “external_plugins” allows me to load a plugin that doesn’t exist in that directory.

**Comment by brianteeman** [comment]
OK so can you please remove the tiny.cloud part of this PR so its just for the plugins and if possible provide a link to a sample plugin for testing purposes

**Comment by ghost** [comment]
I support Brian position on GPL that it goes against our core values.

**Comment by martyf** [comment]
Can do. Did you want a new PR or update this one?

**Comment by brianteeman** [comment]
You can just update this one

**Comment by dgrammatiko** [comment]
@martyf I think you need to move the fieldset/fields to `plugins/editors/tinymce/form/` so that the plugins are available per group not globally. @Fedik am I right on this one?

**Comment by martyf** [comment]
PR updated and removed references to Tiny Cloud.

Example external plugin at:
https://github.com/martyf/tiny-plugin-helloworld

To use:

1. Place the plugin JS file somewhere (could vary based on how the dev wants to set it all up - for example, "/tiny-plugins/helloworld.js"
2. Go to the Joomla 4 Admin, and TinyMCE Plugin settings. On the "External Plugins" tab, create a new entry for the repeatable with:
- **Name:** helloworld
- **Path:** /tiny-plugins/helloworld.js
3. On the main Plugin tab, select Set 0 (i.e. the Set for standard admin users), and on "custom buttons", enter the plugin button - in this case, `helloworld`.
4. Create an Article, and you should see a "Hello Button" in the Tiny toolbar
5. Click the button to trigger the plugin - this is a totally trivial example that demonstrates what Tiny 5 can do with its UI Components library (new with v5) - select an option, and on selecting it, will insert the markup. Totally trivial, but is an example plugin.

**Comment by martyf** [comment]
@dgrammatiko I was thinking that too but then means its very repetitive in its setup - the other thought I had would be to only add the path to the configuration if the button was added to one of the toolbars. That's probably a friendlier setup where the external plugins are configured once, and only called when needed. 

Thoughts?

EDIT
Actually that won’t work - some plugins don’t need a button to be triggered (can be done by a keyboard trigger for example). 

Will have to be moved to the setoptions.xml file and configured per set. 

I’ll get that done and update the PR.

**Comment by martyf** [comment]
Revision made that has moved the External Plugins config to the setoptions.xml file for per-set configuration.

See above for an example (trivial) plugin.

Steps to implement are:

1. Place the plugin JS file somewhere (could vary based on how the dev wants to set it all up - for example, "/tiny-plugins/helloworld.js"
2. Go to the Joomla 4 Admin, and TinyMCE Plugin settings, and select the Set (such as 0 for Admin). Enter a new External Plugin (bottom):
- **Name:** helloworld
- **Path:** /tiny-plugins/helloworld.js
3. In "custom buttons", enter the plugin button - in this case, `helloworld`.
4. Create an Article, and you should see a "Hello Button" in the Tiny toolbar
5. Click the button to trigger the plugin - this is a totally trivial example that demonstrates what Tiny 5 can do with its UI Components library (new with v5) - select an option, and on selecting it, will insert the markup. Totally trivial, but is an example plugin.

**Comment by richard67** [comment]
@martyf After you have removed the cloud stuff, you should update the "Summary of Changes" section at the top of the description of this pull request (PR) accordingly, because it still tells about the cloud. Maybe you should also copy the instructions in your last comment to the "Testing Instructions" section in the description of this PR, so testers (it needs 2 good tests for every PR) can easily find what has been done and how it has to be tested.

**Comment by martyf** [comment]
All done, @richard67

**Comment by richard67** [comment]
@martyf The TinyMCE Editor plugin in J4 has already a field "Custom plugin". Is your External Plugin something else?

**Comment by richard67** [comment]
I mean what's the difference between the existing and the new field?

**Comment by brianteeman** [comment]
@richard67 https://github.com/joomla/joomla-cms/pull/25734#issuecomment-515878705

**Comment by richard67** [comment]
@brianteeman Thanks, I understand now. But it seems the new functionality includes the old one, i.e. with the new option you can do what could be done with the old option and more, so should not the old field be replaced by the new field instead of having both fields? What do you think? Or are the 2 kinds of plugins for TinyMCE different?

**Comment by richard67** [comment]
Ok, I just read all above and understand that with this PR, the plugins can be located anywhere outside the TinyMCE path. Is that the only difference to the existing custom plugins?

**Comment by martyf** [comment]
@richard67 that is correct, external_plugins allows plugins located outside of the Tiny source folder to be loaded.

The Custom Plugins option that exists already is populating the "plugins" config option for Tiny, which is used to load plugins that come packaged with Tiny (but are not necessarily loaded).

Both options need to remain as they do different things - both load plugins, but one loads core where as one loads external.

**Comment by richard67** [comment]
@martyf One final question before I test (or suggest a change, depends on your answer): The path, is that either an absolute or relative path on the server, like I assume? Or can it also be a URL? I've tried with URL, too, because the label "External Plugin URLs" made me think so, but am URL did not work.

**Comment by martyf** [comment]
It can be a web-accessible path (absolute or relative) to a file in the Joomla folder structure (such as a new folder for Tiny plugins, or could even be built and installed as a Joomla plugin too.

It can also be a URL to another server - assuming it is configured to allow cross domain requests.

**Comment by richard67** [comment]
I just saw I made a mistake when testing URL. It works. Was not related to cross-domain.

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on 041ca42dcbcd9c3710459aeaef7aa434057ee54f<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25734">issues.joomla.org/tracker/joomla-cms/25734</a>.</sub>

**Comment by richard67** [comment]
@martyf I've added a section about required documentation changes to the description of this PR. The section is also part of our pull request template.

**Comment by martyf** [comment]
Is there something you wanted me to do to help with that?

**Comment by richard67** [comment]
Well jdoc is a wiki, so if you want, you can sign up there and change the documentation after this PR will be merged. But if not, other jdoc contributors can do that, too.

**Comment by richard67** [comment]
But maybe I am wrong, am not familiar with the docs and if the help is handled differently. Maybe @brianteeman knows that better.

File: plugins/editors/tinymce/tinymce.php [hunk 1]
```diff
@@ -187,6 +187,28 @@ public function onDisplay(
...
+				// if we have a name and path, add it to the list
+				if ($external['name'] != '' && $path != '')
+				{
+					if (substr($path, 0, 1) == '/')
+					{
+						// treat as a local path, so add the root
+						$path = Uri::root() . substr($path, 1);
+					}
+
+					$externalPlugins[ $external['name'] ] = $path;
```
Comment (Quy) [inline plugins/editors/tinymce/tinymce.php]:
```suggestion
					$externalPlugins[$external['name']] = $path;
```

**Comment by wilsonge** [comment]
Thanks!

**Pull request merged**